[run-webkit-tests] Unprefix paths in shards

JonWBedard opened the pull request:
#### e07cfdd0e0cb861502c369c37f8c1012a1bc265f
<pre>
[run-webkit-tests] Unprefix paths in shards
<a href="https://bugs.webkit.org/show_bug.cgi?id=247575">https://bugs.webkit.org/show_bug.cgi?id=247575</a>
rdar://102045852

Reviewed by Elliott Williams.

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
(TestShard.__init__): Make class memeber a packed list of TestInput object.
(TestShard.shorten): Shorten a string based on the shard name.
(TestShard.expand): Expand a string based on the shard name.
(TestShard.pack): Reconstruct a TestInput with mutated paths.
(TestShard.__getstate__): Pack TestShard for serialization.
(TestShard.__setstate__): Unpack TestShard from serialization.

Canonical link: <a href="https://commits.webkit.org/256514@main">https://commits.webkit.org/256514@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c85723c29f1f3008b21a7f8ed797d4cafb7e01a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95979 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5228 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29020 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105538 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165861 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5336 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33990 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88361 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101361 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101640 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3934 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82584 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30978 "Found 1 new test failure: media/video-as-img-output-pts.html (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/99565 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87693 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73811 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39721 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37399 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20562 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4506 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/41995 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/43882 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39822 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -501,9 +502,65 @@ class TestShard(object):
+
+    @property
+    def test_inputs(self):
+        return [self.unpack(input) for input in self._test_inputs]
```
I'm a little concerned about extra overhead to expose this property, since accessing it is now a `N(# of inputs)` operation, and we call it for our `__eq__` implementation. I think we should either treat packing as a "serialization" step and only do it when sending the shard over the wire, or we should make this a callable method e.g. `unpacked_test_inputs()`. If we go with the former, implementing `__getstate__()` and `__setstate()__` seems like the pythonic way to do this.

JonWBedard replied:
I suspect that `__eq__`'s performance is bad even without my change, but using `__getstate__` and `__setstate__` is a good idea...I hadn't used it originally for Python2/3 compatibility, but that seems silly given current support of Python 2

dewei-zhu commented on Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
```diff
@@ -504,6 +505,70 @@ def __init__(self, name, test_inputs):
+            should_run_pixel_test=test_input.should_run_pixel_test,
+        )
+
+    def unpack(self, test_input):
```
It looks like `pack` and `unpack` are almost identical, the only difference is how we operated on path, maybe we can have a helper function to take `self.shorten` and `self.expand` as argument.

JonWBedard committed:
[run-webkit-tests] Unprefix paths in shards
https://bugs.webkit.org/show_bug.cgi?id=247575
rdar://102045852

Reviewed by Elliott Williams.

* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_runner.py:
(TestShard.__init__): Make class memeber a packed list of TestInput object.
(TestShard.shorten): Shorten a string based on the shard name.
(TestShard.expand): Expand a string based on the shard name.
(TestShard.pack): Reconstruct a TestInput with mutated paths.
(TestShard.__getstate__): Pack TestShard for serialization.
(TestShard.__setstate__): Unpack TestShard from serialization.

Canonical link: https://commits.webkit.org/256514@main
